python/ray/air/tests/test_integration_wandb.py

```diff
@@ -345,29 +380,95 @@ def test_set_serializability_result(self, trial):
+        logger.on_trial_complete(0, [], trial)
+        logger.on_experiment_end(trials=[trial])
+        logged = logger.trial_logging_actor_states[trial].logs[0]
         assert logged != "serialization error"
 
-    def test_wandb_logging_actor_api_key(self, ray_start_2_cpus, trial, monkeypatch):
+    def test_wandb_logging_actor_api_key(self, trial, monkeypatch):
         """Tests that the wandb API key get propagated as an environment variable to
         the remote logging actors."""
 
         def mock_run(actor_cls):
             return os.environ.get(WANDB_ENV_VAR)
 
-        monkeypatch.setattr(
-            WandbLoggerCallback, "_logger_actor_cls", _MockWandbLoggingActor
-        )
         monkeypatch.setattr(_MockWandbLoggingActor, "run", mock_run)
 
         logger = WandbLoggerCallback(
             project="test_project", api_key="1234", excludes=["metric2"]
         )
+        logger._logger_actor_cls = _MockWandbLoggingActor
         logger.setup()
         logger.log_trial_start(trial)
         actor_env_var = ray.get(logger._trial_logging_futures[trial])
         assert actor_env_var == "1234"
 
+    def test_wandb_finish(self, trial, tmp_path):
+        """Test that logging actors are cleaned up upon experiment completion."""
+        marker = tmp_path / "hang_marker"
+        marker.write_text("")
+
+        class HangingFinishMockWandbAPI(_MockWandbAPI):
+            def finish(self):
+                while marker.exists():
+                    time.sleep(0.1)
+
+        logger = get_mock_wandb_logger(
+            mock_api_cls=HangingFinishMockWandbAPI,
+            upload_timeout=1.0,
+        )
+        logger.setup()
+        logger.on_trial_start(0, [], trial)
+        logger.on_trial_complete(0, [], trial)
+        # Signalling stop will not cleanup fully due to the hanging finish
+        assert logger._trial_logging_actors
+        marker.unlink()
+        # wandb.finish has ended -> experiment end hook should cleanup actors fully
+        logger.on_experiment_end(trials=[trial])
+        assert not logger._trial_logging_actors
+
+    def test_wandb_kill_hanging_actor(self, trial):
+        """Test that logging actors are killed if exceeding the upload timeout
+        upon experiment completion."""
+
+        class HangingFinishMockWandbAPI(_MockWandbAPI):
+            def finish(self):
+                time.sleep(5)
+
+        logger = get_mock_wandb_logger(
+            mock_api_cls=HangingFinishMockWandbAPI,
+            upload_timeout=0.1,
+        )
+        logger.setup()
+        logger.on_trial_start(0, [], trial)
+        logger.on_trial_complete(0, [], trial)
+        # Signalling stop will not cleanup fully due to the hanging finish
+        assert logger._trial_logging_actors
+        # Experiment end hook should kill actors since upload_timeout < 5
+        logger.on_experiment_end(trials=[trial])
+        assert not logger._trial_logging_actors
```

Comment:
I tried adding an assertion at the end to make sure the actor was killed - but seems like `ray.kill` doesn't happen instantly -- needed to add a sleep of 0.1 seconds in order for a remote method call to fail.

Reply:
Using `gc.collect` before works.